posts/extend-dataframes/index.qmd

```diff
@@ -0,0 +1,283 @@
+  birthdays_reconstruct(data, template)
+}
+```
+
+Information about the generic can be found through the {dplyr} help documentation.
+
+```{r}
+?dplyr_extending
+?dplyr::dplyr_reconstruct
+```
+
+As explained in the help documentation, {dplyr} also uses two base R functions to perform data manipulation. 
+`names<-` (i.e the names setter function) and `[` the one-dimensional subsetting function. We therefore define these methods for our custom class in order for `dplyr_reconstruct()` to work as intended.
+
+```{r}
+`[.birthdays` <- function(x, ...) {
+  out <- NextMethod()
+  birthdays_reconstruct(out, x)
+}
+
+`names<-.birthdays` <- function(x, value) {
+  out <- NextMethod()
+  birthdays_reconstruct(out, x)
+}
+```
+
+This wraps up the need for adding function to perform data manipulation using the reconstruction design outlined above. 
+
+However, there is some final housekeeping to do. In cases when {dplyr} is not a package dependency (either imported or suggested), then the S3 generic `dplyr_reconstruct()` is required to be loaded. In R versions before 3.6.0 -- this also works for R versions later than 3.6.0 -- the generic function needs to be registered. This is done by writing an `.onLoad()` function, typically in a file called zzz.R. This is included in the {ExtendDataFrames} package for illustrative purposes.
+
+```{r}
+.onLoad <- function(libname, pkgname) {
+  s3_register("dplyr::dplyr_reconstruct", "birthdays")
+  invisible()
+}
+```
+
+The `s3_register()` function used in `.onLoad()` also needs to be added to the package and this function is kindly supplied by both {vctrs} and {rlang} unlicensed and thus can be copied into another package. See the [R packages book](https://r-pkgs.org/dependencies-mindset-background.html#sec-dependencies-attach-vs-load) for information about `.onLoad()` and attaching and loading in general.
+
+Since R version 3.6.0 this [S3 generic registration](https://blog.r-project.org/2019/08/19/s3-method-lookup/index.html) happens automatically with `S3Method()` in the package namespace using the {roxygen2} documentation `#' @exportS3Method dplyr::dplyr_reconstruct`.
+
+There is one last option which prevents the hard dependency on a relatively recent R version. Since {roxygen2} version 6.1.0, there is the `@rawNamespace` tag which allows insertion of text into the NAMESPACE file. Using this tag the following code will check the local R version and register the S3 method if equal to or above 3.6.0.
+
+```{r}
+#' @rawNamespace if (getRversion() >= "3.6.0") {
+#'   S3method(pkg::fun, class)
+#' }
+```
+
+Each of the three options for registering S3 methods has different benefits and downsides, so the choice depends on the specific use-case. Over time it may be best to use the most up-to-date methods as packages are usually only maintained for a handful of recent R releases[^1].
+
+The topics discussed in this post have been implemented in the {epiparameter} R package within Epiverse-TRACE. 
```

Comment:
Consider hyperlinking `epiparameter` and Epiverse-TRACE.

Reply:
Resolved in 242917d60b74109afcf7921d9a17c0afc306e9a5